BioSQL/BioSeq.py

```diff
@@ -43,147 +43,45 @@ def __len__(self):
                 self.primary_id, self.start + i, self.start + i + 1
             )
-        if not isinstance(index, slice):
-            raise TypeError("Unexpected index type")
+            return ord(c)
 
-        # Return the (sub)sequence as another DBSeq or Seq object
-        # (see the Seq obect's __getitem__ method)
-        if index.start is None:
-            i = 0
-        else:
-            i = index.start
-        if i < 0:
-            # Map to equavilent positive index
-            if -i > self._length:
-                raise IndexError(i)
-            i = i + self._length
-        elif i >= self._length:
-            # Trivial case, should return empty string!
-            i = self._length
-
-        if index.stop is None:
-            j = self._length
-        else:
-            j = index.stop
-        if j < 0:
-            # Map to equavilent positive index
-            if -j > self._length:
-                raise IndexError(j)
-            j = j + self._length
-        elif j >= self._length:
-            j = self._length
-
-        if i >= j:
-            # Trivial case, empty string.
-            return Seq("")
-        elif index.step is None or index.step == 1:
-            # Easy case - can return a DBSeq with the start and end adjusted
-            return self.__class__(
-                self.primary_id, self.adaptor, None, self.start + i, j - i
+        if step == 1:
+            # Return a _BioSQLSequenceData with the start and end adjusted
+            return _BioSQLSequenceData(
+                self.primary_id, self.adaptor, self.start + start, size
             )
         else:
-            # Tricky.  Will have to create a Seq object because of the stride
+            # Will have to extract the sequence because of the stride
             full = self.adaptor.get_subseq_as_string(
-                self.primary_id, self.start + i, self.start + j
+                self.primary_id, self.start + start, self.start + end
             )
-            return Seq(full[:: index.step])
-
-    def tostring(self):
-        """Return the full sequence as a python string (DEPRECATED).
-
-        You are now encouraged to use str(my_seq) instead of
-        my_seq.tostring().
-        """
-        import warnings
-        from Bio import BiopythonDeprecationWarning
-
-        warnings.warn(
-            "This method is obsolete; please use str(my_seq) "
-            "instead of my_seq.tostring().",
-            BiopythonDeprecationWarning,
-        )
-        return self.adaptor.get_subseq_as_string(
-            self.primary_id, self.start, self.start + self._length
-        )
+            return full[:: step].encode("ASCII")
 
     def __bytes__(self):
         """Return the full sequence as bytes."""
-        return str(self).encode("ASCII")
-
-    def __str__(self):
-        """Return the full sequence as a python string."""
-        return self.adaptor.get_subseq_as_string(
+        sequence = self.adaptor.get_subseq_as_string(
```

Comment:
Is this bytes or unicode?

Reply:
bytes

Reply:
This method ``get_subseq_as_string`` returns the value from the SQL query "SELECT SUBSTR(seq, %s, %s) FROM biosequence WHERE bioentry_id = %s" but it is not clear if this is going to be unicode or bytes.

Prior to your pull request, this was used for ``__str__`` suggesting it was a unicode? Now you say is bytes. Does that mean you've found and fixed another Python 2/3 oversight?

Or is this more complicated in that bytes vs unicode depends on the SQL back end and/or adaptor (e.g. one of the MySQL interfaces, SQLite3, etc)?

Reply:
`sequence` is `str`, which is then transformed to `bytes` on the next line by calling `encode`

Reply:
Clear now, thanks. That final encode wasn't visually close in the diff view.